r-package/grf/vignettes/llf.Rmd

```diff
@@ -0,0 +1,147 @@
+---
+title: "Introduction to local linear forests"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{llf}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+set.seed(123)
+```
+
+```{r setup, warning = FALSE, message = FALSE}
+library(grf)
+library(glmnet)
+library(ggplot2)
+```
+
+This document aims to show how to implement local linear forests (LLF). We begin with the standard use case, walking through parameter choices and method details, and then discuss how to use local linear corrections with larger datasets. 
+
+```{r}
+n <- 600
+p <- 20
+sigma <- sqrt(20)
+
+mu <- function(x){
+  2 * x[1] + x[2] + x[3] * x[4]
+}
+
+X <- matrix(rnorm(n*p, 0, 1), nrow = n)
+Y <- apply(X, FUN = mu, MARGIN = 1) + sigma * rnorm(n)
+
+X.test <- matrix(rnorm(n*p, 0, 1), nrow = n)
+truth = apply(X.test, FUN = mu, MARGIN = 1)
+
+# regression forest predictions
+forest <- regression_forest(X, Y, honesty = TRUE)
+results <- predict(forest, X.test) 
+preds <- results$predictions
+mean((preds - truth)**2)
+```
+
+We can get LLF predictions both from a standard regression forest by specifying linear correction variables, or from a ll_regression_forest object. The parameter linear correction variables gives the variables to use for the final local regression step. This can simply be all variables, or might be a subset. 
+
+```{r}
+# llf predictions from regression_forest
+results.llf <- predict(forest, X.test, linear.correction.variables = 1:p) 
+preds.llf <- results.llf$predictions
+mean((preds.llf - truth)**2)
+
+# llf predictions from ll_regression_forest
+forest <- ll_regression_forest(X, Y, honesty = TRUE)
+results.llf <- predict(forest, X.test) 
+preds.llf <- results.llf$predictions
+mean((preds.llf - truth)**2)
+```
+
+When we perform LLF predictions, we can either do a standard ridge regression (ll.weight.penalty set to FALSE), or scale by the covariance matrix (ll.weight.penalty set to TRUE). This defaults to FALSE.
+
+```{r}
+results.llf.weighted <- predict(forest, X.test, 
+                               ll.weight.penalty = TRUE) 
+preds.llf.weighted <- results.llf.weighted$predictions
+mean((preds.llf.weighted - truth)**2)
+```
+
+The user can choose to specify a ridge regression parameter ll.lambda. This defaults to NULL, and if not modified will be selected by automatic parameter tuning. In general, we recommend letting the forest tune this parameter, or performing your own cross-validation loop. The exception to this would be for very large datasets. 
```

Comment:
```suggestion
The user can choose to specify a ridge regression parameter ll.lambda. When this variable is not set by the user, it will be selected by automatic parameter tuning. In general, we recommend letting the forest tune this parameter, or performing your own cross-validation loop. The exception to this would be for very large datasets. 
```